frontend/lib/src/components/elements/Dialog/Dialog.tsx

```diff
@@ -0,0 +1,142 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, {
+  ReactElement,
+  ReactNode,
+  useEffect,
+  useMemo,
+  useState,
+} from "react"
+
+import { SIZE } from "baseui/modal"
+
+import { EmotionTheme } from "@streamlit/lib/src/theme"
+import Modal, {
+  ModalHeader,
+  ModalBody,
+} from "@streamlit/lib/src/components/shared/Modal"
+import { notNullOrUndefined } from "@streamlit/lib/src/util/utils"
+import { Block as BlockProto } from "@streamlit/lib/src/proto"
+import { LibContext } from "@streamlit/lib/src/components/core/LibContext"
+import ThemeProvider from "@streamlit/lib/src/components/core/ThemeProvider"
+import IsSidebarContext from "@streamlit/lib/src/components/core/IsSidebarContext"
+import { WidgetStateManager } from "@streamlit/lib/src/WidgetStateManager"
+
+import { StyledDialogContent } from "./styled-components"
+
+export interface Props {
+  element: BlockProto.Dialog
+  widgetMgr: WidgetStateManager
+}
+
+type DialogWidth = "small" | "large"
+
+function parseWidthConfig(width: DialogWidth, theme: EmotionTheme): string {
+  if (width === "large") {
+    // this is the same width including padding as the AppView container is using for all inner elements
+    // the padding is added to the ModalBody and subtracted from the total width here
+    // As of writing this: total width=752px (47rem), padding left and right each=24px (1.5rem) => content width = 704px
+    return `calc(${theme.sizes.contentMaxWidth} + 2*${theme.spacing.sm})`
+  }
+
+  return SIZE.default
+}
+
+function isDialogWidth(str: string | null | undefined): str is DialogWidth {
+  return str === "small" || str === "large"
+}
+
+const Dialog: React.FC<React.PropsWithChildren<Props>> = ({
+  element,
+  children,
+  widgetMgr,
+}): ReactElement => {
+  const { title, dismissible, width, isOpen: initialIsOpen } = element
+  const [isOpen, setIsOpen] = useState<boolean>(false)
+
+  useEffect(() => {
+    // release the lock properly when isOpen changed / the component unmounts
+    return () => widgetMgr.releaseDialogLock()
+  }, [widgetMgr])
+
+  useEffect(() => {
+    // Only apply the open state if it was actually set in the proto.
+    if (notNullOrUndefined(initialIsOpen)) {
+      // only open when the lock can be acquired
+      const lockAcquired = widgetMgr.tryAcquireDialogLock()
+      setIsOpen(initialIsOpen && lockAcquired)
+    }
+  }, [initialIsOpen, widgetMgr])
+
+  const { activeTheme } = React.useContext(LibContext)
+  const isInSidebar = React.useContext(IsSidebarContext)
+
+  const ThemedModal = useMemo(() => {
```

Comment:
Making sure I understand, this is to ensure the right theme is going in the dialog and to offset the theme from the sidebar.

Curious if we considered alternatives. My current thought is to consider the event root container similar to `st.toast`. @mayagbarnes might have context here. Seems like dialogs escape the boundaries of the app and are applied on the event root container.

Reply:
Agree with Ken here - this was the intended purpose of events, which we defined as elements occurring outside the normal app layout/flow. `st.toast` and `st.dialog` certainly fit this criteria, so would be curious to understand any context. Using `BlockNodeRenderer` seems a bit counterintuitive to me.

Reply:
Ah I did not know about events / the EventContainer! This is great and seems to work well. One issue I am right now looking into though is that the nested dialog check does not work anymore when using the event-deltagenerator.

Reply:
@kmcgrady @mayagbarnes  
I have looked more into the issue with regards to the [check_nested_element_violation](https://github.com/streamlit/streamlit/pull/8040/files#diff-318e52a9586ec15a31a320ec0558ea3e4efa0456f90a857134c3ddd0908ca697R880).
For context: the function raises an exception when some unintended nesting is detected and is simply a refactor (move to a function) of [these existing lines](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/delta_generator.py#L592-L618) plus the addition of checking for nested dialogs.

This seems to stem from `self._active_dg` [in this line](https://github.com/streamlit/streamlit/pull/8040/files#diff-318e52a9586ec15a31a320ec0558ea3e4efa0456f90a857134c3ddd0908ca697R586) does not return the "correct" delta generator in this case. The active delta generator is returned in these (recursive) property functions: `_active_dg`,` _main_dg` [here](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/delta_generator.py#L307-L331).
The “issue” is that for the `event_dg`, there is always the `main_dg` specified as the parent [here](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/delta_generator.py#L788). So in contrast to having the dialog (or any other container) in the `main_dg`, [this if branch](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/delta_generator.py#L314) in `_active_dg` is not true and the dialog nesting is not detected. If I change
```python
@property
def _active_dg(self) -> DeltaGenerator:
    """Return the DeltaGenerator that's currently 'active'.
    If we are the main DeltaGenerator, and are inside a `with` block that
    creates a container, our active_dg is that container. Otherwise,
    our active_dg is self.
    """
    if self == self._main_dg:
        # We're being invoked via an `st.foo` pattern - use the current
        # `with` dg (aka the top of the stack).
        current_stack = dg_stack.get()
        if len(current_stack) > 1:
            return current_stack[-1]

    # We're being invoked via an `st.sidebar.foo` pattern - ignore the
    # current `with` dg.
    return self
```
to 
```python
@property
def _active_dg(self) -> DeltaGenerator:
    """Return the DeltaGenerator that's currently 'active'.
    If we are the main DeltaGenerator, and are inside a `with` block that
    creates a container, our active_dg is that container. Otherwise,
    our active_dg is self.
    """
    if self == self._main_dg:
        # We're being invoked via an `st.foo` pattern - use the current
        # `with` dg (aka the top of the stack).
        current_stack = dg_stack.get()
        if len(current_stack) > 1:
            return current_stack[-1]
    elif self == event_dg and self._main_dg == main_dg:
        # We're being invoked via an `st.foo` pattern - use the current
        # `with` dg (aka the top of the stack).
        current_stack = dg_stack.get()
        if len(current_stack) > 2:
            return current_stack[-2]

    # We're being invoked via an `st.sidebar.foo` pattern - ignore the
    # current `with` dg.
    return self
```
if works because now `_active_dg` is returning the `event_dg` and the nested check takes the blocks into account as well [here](https://github.com/streamlit/streamlit/pull/8040/files#diff-318e52a9586ec15a31a320ec0558ea3e4efa0456f90a857134c3ddd0908ca697R591).

Alternatively, instead of modifying the `_active_dg` function, the line [here](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/delta_generator.py#L590) could be modified to also account for the `event_dg` path similar to the change made above. Maybe a new function called `_active_event_dg` that follows the logic above for the `event_dg` case could be added that returns an empty set if not in a `event_dg` branch.

For my demo app making the change to `_active_dg` did not seem to have other side-effects, but I assume `_active_dg` is a rather critical function, so I am hestitant about making such a change. 
What do you think?

Reply:
@kmcgrady @mayagbarnes  
I have looked more into the issue with regards to the [check_nested_element_violation](https://github.com/streamlit/streamlit/pull/8040/files#diff-318e52a9586ec15a31a320ec0558ea3e4efa0456f90a857134c3ddd0908ca697R880).
For context: the function raises an exception when some unintended nesting is detected in a delta_generator block path and is simply a refactor (move to a function) of [these existing lines](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/delta_generator.py#L592-L618) plus the addition of checking for nested dialogs.

This seems to stem from `self._active_dg` [in this line](https://github.com/streamlit/streamlit/pull/8040/files#diff-318e52a9586ec15a31a320ec0558ea3e4efa0456f90a857134c3ddd0908ca697R586) does not return the "correct" delta generator in this case. The active delta generator is returned in these (recursive) property functions: `_active_dg`,` _main_dg` [here](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/delta_generator.py#L307-L331).
The “issue” is that for the `event_dg`, there is always the `main_dg` specified as the parent [here](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/delta_generator.py#L788). So in contrast to having the dialog (or any other container) in the `main_dg`, [this if branch](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/delta_generator.py#L314) in `_active_dg` is not true and the dialog nesting is not detected. If I change
```python
@property
def _active_dg(self) -> DeltaGenerator:
    """Return the DeltaGenerator that's currently 'active'.
    If we are the main DeltaGenerator, and are inside a `with` block that
    creates a container, our active_dg is that container. Otherwise,
    our active_dg is self.
    """
    if self == self._main_dg:
        # We're being invoked via an `st.foo` pattern - use the current
        # `with` dg (aka the top of the stack).
        current_stack = dg_stack.get()
        if len(current_stack) > 1:
            return current_stack[-1]

    # We're being invoked via an `st.sidebar.foo` pattern - ignore the
    # current `with` dg.
    return self
```
to 
```python
@property
def _active_dg(self) -> DeltaGenerator:
    """Return the DeltaGenerator that's currently 'active'.
    If we are the main DeltaGenerator, and are inside a `with` block that
    creates a container, our active_dg is that container. Otherwise,
    our active_dg is self.
    """
    if self == self._main_dg:
        # We're being invoked via an `st.foo` pattern - use the current
        # `with` dg (aka the top of the stack).
        current_stack = dg_stack.get()
        if len(current_stack) > 1:
            return current_stack[-1]
    elif self == event_dg and self._main_dg == main_dg:
        # We're being invoked via an `st.foo` pattern - use the current
        # `with` dg (aka the top of the stack).
        current_stack = dg_stack.get()
        if len(current_stack) > 2:
            return current_stack[-2]

    # We're being invoked via an `st.sidebar.foo` pattern - ignore the
    # current `with` dg.
    return self
```
if works because now `_active_dg` is returning the `event_dg` and the nested check takes the blocks into account as well [here](https://github.com/streamlit/streamlit/pull/8040/files#diff-318e52a9586ec15a31a320ec0558ea3e4efa0456f90a857134c3ddd0908ca697R591).

Alternatively, instead of modifying the `_active_dg` function, the line [here](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/delta_generator.py#L590) could be modified to also account for the `event_dg` path similar to the change made above. Maybe a new function called `_active_event_dg` that follows the logic above for the `event_dg` case could be added that returns an empty set if not in a `event_dg` branch.

For my demo app making the change to `_active_dg` did not seem to have other side-effects, but I assume `_active_dg` is a rather critical function, so I am hestitant about making such a change. 
What do you think?

Reply:
After some offline discussion, for now we ended up having a distinct function in the dialog-decorator file and do a custom dg-stack check there. @kmcgrady put a brief summary [in this comment](https://github.com/streamlit/streamlit/pull/8040#discussion_r1550565750).